src/main/java/io/airlift/slice/SliceUtf8.java

```diff
@@ -0,0 +1,810 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.airlift.slice;
+
+import static io.airlift.slice.Preconditions.checkArgument;
+import static io.airlift.slice.Preconditions.checkPositionIndex;
+import static io.airlift.slice.Preconditions.checkPositionIndexes;
+import static java.lang.Character.MAX_CODE_POINT;
+import static java.lang.Integer.toHexString;
+
+/**
+ * Utility methods for UTF-8 encoded slices.
+ */
+public final class SliceUtf8
+{
+    private static final int REPLACEMENT_CODE_POINT = 0xFFFD;
+
+    private SliceUtf8() {}
+
+    private static final int TOP_MASK32 = 0x80808080;
+    private static final long TOP_MASK64 = 0x8080808080808080L;
+
+    private static final int[] LOWER_CODE_POINTS;
+    private static final int[] UPPER_CODE_POINTS;
+    private static final boolean[] WHITESPACE_CODE_POINTS;
+
+    static {
+        LOWER_CODE_POINTS = new int[MAX_CODE_POINT];
+        UPPER_CODE_POINTS = new int[MAX_CODE_POINT];
+        WHITESPACE_CODE_POINTS = new boolean[MAX_CODE_POINT];
+        for (int codePoint = 0; codePoint < MAX_CODE_POINT; codePoint++) {
+            int type = Character.getType(codePoint);
+            if (type != Character.UNASSIGNED && type != Character.SURROGATE) {
```

Comment:
UNASSIGNED are valid code points. You can run them through `UTF_8.encode(UTF_8.decode(` in Java and see they don't get replaced with U+FFFD